src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs

```diff
@@ -69,16 +89,24 @@ public int Take()
             }
 
             var index = _index;
-            var wasLastBlock = block.Next == null;
 
             if (index < block.End)
             {
                 _index = index + 1;
                 return block.Array[index];
             }
 
+            return TakeMultiBlock();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private int TakeMultiBlock()
+        {
+            var block = _block;
+            var wasLastBlock = block.Next == null;
```

Comment:
Basically. I don't think TakeMultiBlock() should use do/while though because the way Take() is currently written `block.Next` could be null.

The comment seems good though.